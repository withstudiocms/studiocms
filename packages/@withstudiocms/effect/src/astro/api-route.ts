import type { APIContext, APIRoute } from 'astro';
import { runEffect } from '../index.js';
import type {
	EffectAPIRouteHandler,
	EffectRouteOptions,
	ExtractDefinedRoutes,
	RouteHandlerConfig,
	RouteHandlers,
} from './types.js';
import { applyCors, validateRequest } from './utils/api-route.js';

/**
 * Defines an API route handler for Astro using an effect-based approach.
 *
 * @param context - The API context provided by Astro for the current request.
 * @returns A function that takes an `EffectAPIRouteHandler`, executes it with the provided context,
 *          and returns a `Promise<Response>` representing the HTTP response.
 *
 * @example
 * ```typescript
 * export const GET = defineAPIRoute(context)(async (ctx) => {
 *   // Your effectful logic here
 *   return new Response("Hello, world!");
 * });
 * ```
 *
 * @deprecated Use `createEffectRoute` instead for better clarity and consistency.
 */
export const defineAPIRoute =
	(context: APIContext) =>
	async (fn: EffectAPIRouteHandler): Promise<Response> =>
		await runEffect(fn(context));

/**
 * Creates an Astro API route handler that executes the provided effectful handler function.
 *
 * @param fn - The effectful API route handler function to be executed. It should accept an `APIContext` and return an effect.
 * @returns An Astro `APIRoute` handler that runs the effect and returns a `Response`.
 */
export const createEffectAPIRoute = (fn: EffectAPIRouteHandler): APIRoute => {
	return async (context: APIContext): Promise<Response> => {
		return await runEffect(fn(context));
	};
};

/**
 * Wraps an API route handler with effectful middleware, including CORS, validation, timeout, and custom hooks.
 *
 * @param fn - The main effectful API route handler function to execute.
 * @param options - Optional configuration for the route, including CORS, validation, timeout, and middleware hooks.
 * @returns An `APIRoute` function compatible with Astro's API routes.
 *
 * @remarks
 * This utility provides a standardized way to handle API routes with common concerns such as:
 * - Applying CORS headers and handling preflight requests.
 * - Running pre-effect middleware (`onBeforeEffect`).
 * - Validating incoming requests (`validate`).
 * - Enforcing a timeout for the effect execution.
 * - Running post-success (`onSuccess`) and error (`onError`) middleware.
 * - Ensuring CORS headers are present on all responses, including errors.
 *
 * @example
 * ```typescript
 * export const POST = withEffect(async (context) => {
 *   // Your effect logic here
 * }, {
 *   cors: { origin: '*' },
 *   validate: { body: z.object({ name: z.string() }) },
 *   timeout: 5000,
 *   onSuccess: async (response, context) => response,
 *   onError: async (error, context) => new Response('Error', { status: 500 }),
 * });
 * ```
 */
export const withEffectAPI = (
	fn: EffectAPIRouteHandler,
	options: EffectRouteOptions = {}
): APIRoute => {
	return async (context: APIContext): Promise<Response> => {
		try {
			// Apply CORS headers if configured
			const corsHeaders = applyCors(context, options.cors);

			// Handle preflight requests
			if (context.request.method === 'OPTIONS') {
				return new Response(null, {
					status: 204,
					headers: corsHeaders,
				});
			}

			// Apply pre-effect middleware
			let processedContext = context;
			if (options.onBeforeEffect) {
				processedContext = await options.onBeforeEffect(context);
			}

			// Validate request if validators provided
			if (options.validate) {
				const validationError = await validateRequest(processedContext, options.validate);
				if (validationError) {
					const errorResponse = new Response(
						JSON.stringify({ error: 'Validation failed', details: validationError }),
						{ status: 400, headers: { 'Content-Type': 'application/json', ...corsHeaders } }
					);
					return errorResponse;
				}
			}

			// Execute the effect with timeout if specified
			let effectPromise = runEffect(fn(processedContext));
			if (options.timeout) {
				effectPromise = Promise.race([
					effectPromise,
					new Promise<never>((_, reject) =>
						setTimeout(() => reject(new Error('Request timeout')), options.timeout)
					),
				]);
			}

			const response = await effectPromise;

			// Add CORS headers to response
			if (corsHeaders && Object.keys(corsHeaders).length > 0) {
				Object.entries(corsHeaders).forEach(([key, value]) => {
					response.headers.set(key, value);
				});
			}

			// Apply success middleware
			return options.onSuccess ? await options.onSuccess(response, processedContext) : response;
		} catch (error) {
			const corsHeaders = applyCors(context, options.cors);

			if (options.onError) {
				const errorResponse = await options.onError(error, context);
				// Ensure CORS headers on error responses too
				Object.entries(corsHeaders).forEach(([key, value]) => {
					errorResponse.headers.set(key, value);
				});
				return errorResponse;
			}

			// Default error handling with CORS
			return new Response(
				JSON.stringify({
					error: error instanceof Error ? error.message : 'Internal Server Error',
				}),
				{
					status: error instanceof Error && error.message === 'Request timeout' ? 408 : 500,
					headers: {
						'Content-Type': 'application/json',
						...corsHeaders,
					},
				}
			);
		}
	};
};

/**
 * Creates API route handlers with effect support for each HTTP method provided.
 *
 * This function takes a set of route handlers and an optional configuration object,
 * then returns an object mapping HTTP methods to their corresponding API route handlers,
 * each wrapped with effect middleware and merged configuration.
 *
 * The return type is precisely typed to only include the methods that were provided
 * in the handlers parameter, eliminating undefined from the union types.
 *
 * @param handlers - An object containing route handlers keyed by HTTP method (e.g., 'GET', 'POST').
 * @param config - Optional global configuration for all handlers. Can include method-specific overrides via the `methods` property.
 * @returns An object mapping each provided HTTP method to its configured API route handler.
 */
export const createEffectAPIRoutes = <T extends RouteHandlers>(
	handlers: T,
	config: RouteHandlerConfig = {}
): ExtractDefinedRoutes<T> => {
	const routes = {} as ExtractDefinedRoutes<T>;

	for (const [method, handler] of Object.entries(handlers)) {
		if (handler) {
			const methodKey = method as keyof RouteHandlers;

			// Merge global config with method-specific overrides
			const methodConfig = {
				...config,
				...(config.methods?.[methodKey] || {}),
			};

			// Remove the methods property from the final config
			const { methods: _droppedMethods, ...finalConfig } = methodConfig;

			// TypeScript knows this is safe because we checked `if (handler)` above
			// biome-ignore lint/suspicious/noExplicitAny: `routes` is typed as `ExtractDefinedRoutes<T>`
			(routes as any)[methodKey] = withEffectAPI(handler as EffectAPIRouteHandler, finalConfig);
		}
	}

	return routes;
};

/**
 * A builder class for defining API route handlers with optional per-method and global configuration.
 *
 * Provides a fluent interface for registering HTTP method handlers (`get`, `post`, `put`, `delete`, `patch`, `all`)
 * and setting global or per-method options. Use `build()` to generate the final route configuration.
 *
 * Each method returns a new builder type that accumulates the added methods, ensuring precise typing
 * in the final built routes object.
 *
 * @example
 * ```typescript
 * const routes = new EffectRouteBuilder()
 *   .withGlobalConfig({ cors: { origin: true } })
 *   .get(getHandler, { timeout: 3000 })
 *   .post(postHandler)
 *   .build();
 * // routes has type: { GET: APIRoute, POST: APIRoute } (no undefined!)
 * ```
 */

// biome-ignore lint/complexity/noBannedTypes: `T` is used to accumulate handlers
export class EffectAPIRouteBuilder<T extends Partial<RouteHandlers> = {}> {
	private handlers: T;
	private config: RouteHandlerConfig = {};

	constructor(handlers: T = {} as T) {
		this.handlers = handlers;
	}

	// Set global configuration
	public withGlobalConfig(config: EffectRouteOptions): EffectAPIRouteBuilder<T> {
		this.config = { ...this.config, ...config };
		return this;
	}

	private addHandler<K extends keyof RouteHandlers, U extends EffectAPIRouteHandler>(
		method: K,
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & Record<K, U>> {
		const newHandlers = { ...this.handlers, [method]: handler } as T & Record<K, U>;
		const newBuilder = new EffectAPIRouteBuilder(newHandlers);
		newBuilder.config = { ...this.config };

		if (options) {
			newBuilder.config.methods = { ...this.config.methods, [method]: options };
		}
		return newBuilder;
	}

	// Add handlers with fluent interface - each method returns a new type that includes the added method
	public get<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { GET: U }> {
		return this.addHandler('GET', handler, options) as EffectAPIRouteBuilder<T & { GET: U }>;
	}

	public post<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { POST: U }> {
		return this.addHandler('POST', handler, options) as EffectAPIRouteBuilder<T & { POST: U }>;
	}

	public put<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { PUT: U }> {
		return this.addHandler('PUT', handler, options) as EffectAPIRouteBuilder<T & { PUT: U }>;
	}

	public delete<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { DELETE: U }> {
		return this.addHandler('DELETE', handler, options) as EffectAPIRouteBuilder<T & { DELETE: U }>;
	}

	public patch<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { PATCH: U }> {
		return this.addHandler('PATCH', handler, options) as EffectAPIRouteBuilder<T & { PATCH: U }>;
	}

	public all<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { ALL: U }> {
		return this.addHandler('ALL', handler, options) as EffectAPIRouteBuilder<T & { ALL: U }>;
	}

	// Build the final routes with precise typing
	public build(): ExtractDefinedRoutes<T> {
		return createEffectAPIRoutes(this.handlers, this.config);
	}
}
