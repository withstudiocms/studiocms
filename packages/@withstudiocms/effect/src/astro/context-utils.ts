import type { APIContext } from 'astro';
import { Effect, Schema } from '../effect.js';

/**
 * Reads and parses the JSON body from an API request within the given context.
 *
 * @template A - The expected type of the parsed JSON object.
 * @param context - An object containing the API request.
 * @returns An `Effect` that resolves with the parsed JSON object of type `A`, or fails with an `Error` if parsing fails.
 */
export const readAPIContextJson = <A>({ request }: APIContext): Effect.Effect<A, Error, never> =>
	Effect.tryPromise({
		try: () => request.json(),
		catch: () => new Error('Failed to parse JSON from Request'),
	});

/**
 * Reads and parses the JSON body from an API request within the given `APIContext`.
 * Optionally validates and decodes the parsed JSON using a provided `Schema`.
 *
 * @typeParam A - The expected type of the decoded JSON object.
 * @typeParam E - The error type that may be returned by the schema decoder.
 * @typeParam R - The environment type required by the schema decoder.
 * @param context - The API context containing the request to read from.
 * @param schema - Optional schema to validate and decode the parsed JSON.
 * @returns An `Effect` that yields the decoded JSON object of type `A`, or throws an error if parsing or decoding fails.
 */
export const parseAPIContextJson = <A, E, R>(
	context: APIContext,
	schema?: Schema.Schema<A, E, R>
): Effect.Effect<A, Error, R> =>
	Effect.gen(function* () {
		const json = yield* readAPIContextJson<A>(context);
		if (schema) return yield* Schema.decodeUnknown(schema)(json);
		return json as A;
	}).pipe(
		Effect.catchAll((error) => {
			console.error('Failed to read JSON:', error);
			return Effect.fail(new Error('Failed to read JSON'));
		})
	);

/**
 * Reads and parses form data from the API context's request object.
 *
 * Wraps the asynchronous `request.formData()` call in an `Effect.tryPromise`,
 * providing error handling if parsing fails.
 *
 * @param context - The API context containing the request object.
 * @returns An `Effect` that resolves to the parsed form data, or an error if parsing fails.
 */
export const readAPIContextFormData = ({
	request,
}: APIContext): Effect.Effect<FormData, Error, never> =>
	Effect.tryPromise({
		try: () => request.formData(),
		catch: () => new Error('Failed to parse formData from Request'),
	});

/**
 * Parses form data from an `APIContext` into an object, optionally validating and decoding it using a provided schema.
 *
 * @template A - The type of the parsed object.
 * @template E - The error type for schema decoding.
 * @template R - The environment type required by the effect.
 * @param context - The API context containing the form data to parse.
 * @param schema - Optional schema to validate and decode the form data.
 * @returns An `Effect` that yields the parsed object of type `A`, or fails with an `Error` if parsing or decoding fails.
 */
export const parseAPIContextFormDataToObject = <A, E, R>(
	context: APIContext,
	schema?: Schema.Schema<A, E, R>
): Effect.Effect<A, Error, R> =>
	Effect.gen(function* () {
		const formData = yield* readAPIContextFormData(context);
		const entries = Object.fromEntries(formData.entries());

		if (schema) return yield* Schema.decodeUnknown(schema)(entries);

		return entries as A;
	}).pipe(
		Effect.catchAll((error) => {
			console.error('Failed to read form data:', error);
			return Effect.fail(new Error('Failed to read form data'));
		})
	);

/**
 * Attempts to retrieve a string value from a FormData entry by key.
 * Returns an Effect that resolves to the string value if present and of type string,
 * or `null` if the entry is not a string.
 *
 * @param formData - The FormData object to extract the value from.
 * @param key - The key of the entry to retrieve.
 * @returns An Effect that resolves to the string value or `null`.
 */
export const parseFormDataEntryToString = (
	formData: FormData,
	key: string
): Effect.Effect<string | null, Error, never> =>
	Effect.try({
		try: () => {
			const value = formData.get(key);
			if (typeof value !== 'string') {
				return null;
			}
			return value;
		},
		catch: () => new Error(`Failed to parse FormData entry for key: ${key}`),
	});
