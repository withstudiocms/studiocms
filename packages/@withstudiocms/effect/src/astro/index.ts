export * from './api-route.js';
export * from './context-utils.js';
export * from './middleware.js';
export * from './response-helpers.js';
export * from './types.js';
