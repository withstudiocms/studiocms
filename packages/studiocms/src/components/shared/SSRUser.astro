---
import { Icon } from 'studiocms:ui/components/icon';

const heroIconsUserDataUri =
	'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAADSklEQVR4AdSWS2xMYRTHjQXRRupRKVWrJhrxSCRigTQpYUHq0VYsarwi8RgJ0rCwFJEQEQ1psZFQXWgGRZB6NdFalkSFFRutV1UJNhLj9+dcOubeud8dC2lzfvd/7vedc77TO999DB/2n/+GZgOpVCoOSegx5K/N5WJGugIsNhHusNBZqIJiQ/455tqgiDFni9QAVZthIVyBBTAWJsBKeAOL4Tw4m3MD/GebqVoBV2Ox2ArohAHog1bGZ0MfLCJ2A+pkzg1QrRpkB3X4G5p4xdhOkOknkYYSpYE5Vu2pqZ/ctUFdDXOzS5QGvlkplxyXmJ/lnAOJfgSyMh0C0AbVVJcOLkRp4JQV3GeaJmy8SQzUg6xRBxecG2CTaaffpGgli12DeVAAhaDb8CFzhXCZ2Ouokzk3YNU2ordhGXTCALyDS6AH0C10CzhbpAb4z16DHjZxVtCivWgPJCHO3BJ4i+9skRrwqrJIE1TBZCiBGmjy5qNoTg1EWSAsNlIDbLZi2AEt0A2fDPkX8BOguyFs3d/zTg1QNB+OkPUcjkMNTIfRhvzV+CfgBbGHYBR+qIU2QCG97dqpVAcjoRuOgpqYggr5GnvCuWL2ou3kjkOzWtYGKFBAdgfoPfAR1S02iw1XB0l4achXgzOJ2QaKnYt2WA1cfwtsgETNtZA2FXR/l7LYaUhx7muag5NMloI+XKahzdSKob6mRXwnGEyA7nm949dQ+D3nTmax2hN6SC0lSVcOyTTfBuhYl36/hddS8IP5zmI56y3hADW1Ye30j/g2wLQ21Ri0i0JtaE5G7g0SH8N40PsCSbegBiot7J7pv4hXw6uZViuogRkWpdvP3JzFq6E7JKNIUAP63FbwJn67KsjTSRSUA/o2XGd5Xk07/SVBDei7/zshq0Bvun6KtcIeqIUKKAM9IfNQ+RrTnGL07dBvufrtVesM5xnm2wCbZyuRer9vR+/DCFgOh0FvPX18PsP/DF9AvsY0pxjFKke5up2LqLmLuAzzbUBRJOh7vxEt57wEVKABvQgPQO+Fr6iQrzHNKWY343pNl5PfAHqWMJRpgQ0MDqVAL9RDAqphPujJmI8K+RrTnGKOMa6PlcFlfH2nBnwzHQfDwn4AAAD//6qWhy8AAAAGSURBVAMAJXQ0UKI3Vu0AAAAASUVORK5CYII=';

interface Props {
	/**
	 * The name of the user.
	 */
	name: string;
	/**
	 * The description of the user. Could be a role, a handle, etc.
	 */
	description: string;
	/**
	 * The avatar of the user. Either a URL to an image or an imported image.
	 */
	avatar?: string;
	/**
	 * Additional classes to apply to the user container.
	 */
	class?: string;
	/**
	 * The loading strategy for the image. Defaults to `lazy`.
	 */
	loading?: 'eager' | 'lazy';

	/**
	 * The timeout delay for loading the avatar image. Defaults to 5000ms.
	 */
	timeoutDelay?: number;

	id: string;
}

const {
	name,
	description,
	avatar,
	class: className,
	loading = 'lazy',
	timeoutDelay = 3000,
	id,
} = Astro.props;
---

<div class="sui-user-container" class:list={[className]} id={id}>
    <div class="sui-avatar-container">
        {
			avatar ? (
				<studiocms-avatar
					data-avatar-url={avatar}
					data-avatar-timeout={timeoutDelay}
					data-avatar-fallback={heroIconsUserDataUri}
					data-avatar-name={name}
					data-avatar-loading={loading}
				>
					<img
						src={heroIconsUserDataUri}
						width={64}
						height={64}
						class="sui-avatar-img"
						alt={name}
						loading={loading}
						decoding="async"
						referrerpolicy="no-referrer"
					/>
				</studiocms-avatar>
			) : (
                <Icon
                    name="heroicons:user"
                    width={24}
                    height={24}
                    role="img"
                    aria-label={`Placeholder avatar for ${name}`}
                />
            )
        }
    </div>
    <div class="sui-text-content">
        <span class="sui-name">{name}</span>
        <span class="sui-description">{description}</span>
    </div>
</div>

<script>
	class StudioCMSAvatar extends HTMLElement {
		connectedCallback() {
			const avatarUrl = this.dataset.avatarUrl;
			const timeoutDelay = parseInt(this.dataset.avatarTimeout || '3000', 10);
			const fallbackDataUri = this.dataset.avatarFallback;
			const avatarName = this.dataset.avatarName || 'User';

			if (!avatarUrl) return;

			const img = this.querySelector('img');
			if (!img) return;

			this.loadAvatar(img, avatarUrl, timeoutDelay, fallbackDataUri, avatarName);
		}

		isValidAvatarUrl(url: string): boolean {
			try {
				const urlObj = new URL(url);

				const validProtocols = ['https:', 'data:'];
				if (!validProtocols.includes(urlObj.protocol)) {
					return false;
				}

				const isSvg =
					url.toLowerCase().endsWith('.svg') ||
					url.toLowerCase().includes('.svg?') ||
					url.toLowerCase().includes('image/svg');

				if (isSvg) {
					return false;
				}

				return true;
			} catch {
				return false;
			}
		}

		async verifyAvatarUrl(url: string, timeoutMs: number): Promise<boolean> {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), timeoutMs);
			try {
				const response = await fetch(url, {
					method: 'HEAD',
					signal: controller.signal,
					cache: 'no-cache',
				});

				return response.ok;
			} catch (error) {
				return false;
 	        } finally {
                clearTimeout(timeoutId);
			}
		}

		showFallbackIcon(img: HTMLImageElement, name: string): void {
			const container = img.parentElement;
			if (!container) return;

			const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
			svg.setAttribute('xmlns', 'http://www.w3.org/2000/svg');
			svg.setAttribute('width', '24');
			svg.setAttribute('height', '24');
			svg.setAttribute('viewBox', '0 0 24 24');
			svg.setAttribute('fill', 'none');
			svg.setAttribute('stroke', 'currentColor');
			svg.setAttribute('stroke-width', '2');
			svg.setAttribute('stroke-linecap', 'round');
			svg.setAttribute('stroke-linejoin', 'round');
			svg.setAttribute('role', 'img');
			svg.setAttribute('aria-label', `Placeholder avatar for ${name}`);
			const path1 = document.createElementNS('http://www.w3.org/2000/svg', 'path');
			path1.setAttribute('d', 'M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2');
	
			const circle = document.createElementNS('http://www.w3.org/2000/svg', 'circle');
			circle.setAttribute('cx', '12');
			circle.setAttribute('cy', '7');
			circle.setAttribute('r', '4');

			svg.appendChild(path1);
			svg.appendChild(circle);

			container.removeChild(img);
			container.appendChild(svg);
		}

		async loadAvatar(
			img: HTMLImageElement,
			url: string,
			timeoutMs: number,
			fallbackDataUri: string | undefined,
			name: string
		): Promise<void> {
			if (!this.isValidAvatarUrl(url)) {
				this.showFallbackIcon(img, name);
				return;
			}

			if (url.startsWith('data:')) {
				img.src = url;
				img.classList.add('sui-avatar-loaded');
				return;
			}

			const isAccessible = await this.verifyAvatarUrl(url, timeoutMs);

			if (!isAccessible) {
				if (fallbackDataUri && img.src !== fallbackDataUri) {
					img.src = fallbackDataUri;
				} else {
					this.showFallbackIcon(img, name);
				}
				return;
			}
			img.onload = () => {
				img.classList.add('sui-avatar-loaded');
			};

			img.onerror = () => {
				if (fallbackDataUri && img.src !== fallbackDataUri) {
					img.src = fallbackDataUri;
				} else {
					this.showFallbackIcon(img, name);
				}
			};
			img.src = url;
		}
	}
	if (!customElements.get('studiocms-avatar')) {
		customElements.define('studiocms-avatar', StudioCMSAvatar);
	}
</script>

<style>
    .sui-user-container {
        display: flex;
        flex-direction: row;
        align-items: center;
        gap: 1rem;
    }
    .sui-avatar-container {
        width: 2.5rem;
        height: 2.5rem;
        background-color: var(--default-base);
        border-radius: var(--radius-full);
        display: flex;
        align-items: center;
        justify-content: center;
        overflow: hidden;
        border: 1px solid var(--border);
    }
    .sui-avatar-img {
        width: 100%;
        height: auto;
        display: flex;
        align-items: center;
        justify-content: center;
		opacity: 0.5;
		transition: opacity 0.2s ease-in;
    }
	.sui-avatar-img.sui-avatar-loaded {
		opacity: 1;
	}
    .sui-text-content {
        display: flex;
        flex-direction: column;
        gap: 0.125rem;
    }
    .sui-name {
        font-size: 1em;
        font-weight: 600;
    }
    .sui-description {
        font-size: 0.875em;
        font-weight: 400;
        color: var(--text-muted);
    }
</style>
