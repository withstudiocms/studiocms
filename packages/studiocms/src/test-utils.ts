import type { AstroIntegration, AstroIntegrationLogger } from 'astro';
import type { StudioCMSPlugin } from './plugins.js';
import type {
	SCMSAuthServiceFnOpts,
	SCMSDashboardFnOpts,
	SCMSFrontendFnOpts,
	SCMSImageServiceFnOpts,
	SCMSRenderingFnOpts,
	SCMSSiteMapFnOpts,
} from './schemas/index.js';

type HookRun<T> = { hasHook: boolean; hookResults: T };
export interface PluginHookResults {
	astroConfig: HookRun<{ integrations: AstroIntegration[] }>;
	studiocmsConfig: HookRun<{
		authService: Partial<SCMSAuthServiceFnOpts>;
		dashboard: Partial<SCMSDashboardFnOpts>;
		frontend: Partial<SCMSFrontendFnOpts>;
		imageService: Partial<SCMSImageServiceFnOpts>;
		rendering: Partial<SCMSRenderingFnOpts>;
		sitemap: Partial<SCMSSiteMapFnOpts>;
	}>;
}
/**
 * Utility class for testing StudioCMS plugins by simulating hook execution and collecting results.
 *
 * The `StudioCMSPluginTester` provides methods to:
 * - Create mock loggers for use in plugin hooks.
 * - Execute the `studiocms:astro:config` and `studiocms:config:setup` hooks, if present, on a given plugin.
 * - Collect and return integration and configuration data set by these hooks.
 * - Retrieve basic plugin metadata.
 *
 * @remarks
 * This class is intended for use in test environments to facilitate inspection and validation of plugin behavior.
 *
 * @example
 * ```typescript
 * const tester = new StudioCMSPluginTester(myPlugin);
 * const info = tester.getPluginInfo();
 * const hookResults = await tester.getHookResults();
 * ```
 */
export class StudioCMSPluginTester {
	private plugin: StudioCMSPlugin;
	private readonly injectedLogger?: AstroIntegrationLogger;

	constructor(plugin: StudioCMSPlugin, logger?: AstroIntegrationLogger) {
		this.plugin = plugin;
		this.injectedLogger = logger;
	}

	/**
	 * Creates a mock logger object for testing purposes.
	 *
	 * The returned logger implements the `AstroIntegrationLogger` interface and provides
	 * stubbed logging methods (`info`, `warn`, `error`, `debug`) that delegate to the corresponding
	 * `console` methods. It also includes a `fork` method, which returns a new mock logger
	 * instance with an attached label for testing logger forking behavior.
	 *
	 * @returns {AstroIntegrationLogger} A mock logger instance suitable for use in tests.
	 */
	private createMockLogger(): AstroIntegrationLogger {
		if (this.injectedLogger) return this.injectedLogger;
		// biome-ignore lint/suspicious/noExplicitAny: this is fine
		const logger: Record<string, any> = {
			info: console.log,
			warn: console.warn,
			error: console.error,
			debug: console.debug,
			fork: (label: string) => {
				// Each fork returns a new mock logger with the label attached for testing
				const forked = this.createMockLogger();
				forked.label = label;
				return forked;
			},
		};
		return logger as unknown as AstroIntegrationLogger;
	}

	/**
	 * Executes the 'studiocms:astro:config' hook if it exists on the plugin,
	 * providing a mock logger and a method to collect Astro integrations.
	 *
	 * @returns A promise that resolves to an object containing the collected integrations.
	 *
	 * @remarks
	 * This method checks if the plugin defines a 'studiocms:astro:config' hook as a function.
	 * If so, it invokes the hook with a mock logger and an `addIntegrations` callback,
	 * which accumulates integrations into an array. The collected integrations are then
	 * returned in an object.
	 */
	private async runAstroConfigHook() {
		const integrations: AstroIntegration[] = [];
		if (
			this.plugin.hooks['studiocms:astro:config'] &&
			typeof this.plugin.hooks['studiocms:astro:config'] === 'function'
		) {
			await this.plugin.hooks['studiocms:astro:config']({
				logger: this.createMockLogger(),
				addIntegrations: (newIntegrations) => {
					const toAdd = Array.isArray(newIntegrations) ? newIntegrations : [newIntegrations];
					integrations.push(...toAdd);
				},
			});
		}
		return { integrations };
	}

	/**
	 * Executes the `studiocms:config:setup` hook on the current plugin, if available,
	 * and collects configuration options set by the hook into partial option objects.
	 *
	 * This method mocks the hook's context by providing setter functions for various
	 * StudioCMS configuration aspects, such as authentication, dashboard, frontend,
	 * image service, rendering, and sitemap. Each setter captures the values provided
	 * by the hook and stores them in corresponding partial option objects.
	 *
	 * @returns A promise that resolves to an object containing the collected configuration
	 *          options for authentication service, dashboard, frontend, image service,
	 *          rendering, and sitemap.
	 *
	 * @remarks
	 * This utility is primarily intended for testing purposes, allowing inspection of
	 * configuration values set by a plugin's `studiocms:config:setup` hook.
	 */
	private async runStudioCMSConfigHook() {
		const authService: Partial<SCMSAuthServiceFnOpts> = {};
		const dashboard: Partial<SCMSDashboardFnOpts> = {};
		const frontend: Partial<SCMSFrontendFnOpts> = {};
		const imageService: Partial<SCMSImageServiceFnOpts> = {};
		const rendering: Partial<SCMSRenderingFnOpts> = {};
		const sitemap: Partial<SCMSSiteMapFnOpts> = {};

		if (
			this.plugin.hooks['studiocms:config:setup'] &&
			typeof this.plugin.hooks['studiocms:config:setup'] === 'function'
		) {
			await this.plugin.hooks['studiocms:config:setup']({
				logger: this.createMockLogger(),
				setAuthService: ({ oAuthProvider }) => {
					if (oAuthProvider) {
						authService.oAuthProvider = oAuthProvider;
					}
				},
				setDashboard: ({ dashboardGridItems, dashboardPages }) => {
					if (dashboardGridItems) {
						dashboard.dashboardGridItems = dashboardGridItems;
					}
					if (dashboardPages) {
						dashboard.dashboardPages = dashboardPages;
					}
				},
				setFrontend: ({ frontendNavigationLinks }) => {
					if (frontendNavigationLinks) {
						frontend.frontendNavigationLinks = frontendNavigationLinks;
					}
				},
				setImageService: ({ imageService: imgService }) => {
					if (imgService) {
						imageService.imageService = imgService;
					}
				},
				setRendering: ({ pageTypes }) => {
					if (pageTypes) {
						rendering.pageTypes = pageTypes;
					}
				},
				setSitemap: ({ sitemaps, triggerSitemap }) => {
					if (sitemaps) {
						sitemap.sitemaps = sitemaps;
					}
					sitemap.triggerSitemap = triggerSitemap;
				},
			});
		}

		return {
			authService,
			dashboard,
			frontend,
			imageService,
			rendering,
			sitemap,
		};
	}

	/**
	 * Retrieves information about the current plugin.
	 *
	 * @returns An object containing the plugin's identifier, name, minimum required StudioCMS version, and dependencies.
	 */
	public getPluginInfo() {
		return {
			identifier: this.plugin.identifier,
			name: this.plugin.name,
			studiocmsMinimumVersion: this.plugin.studiocmsMinimumVersion,
			requires: this.plugin.requires,
		};
	}

	/**
	 * Asynchronously retrieves the results of configured plugin hooks.
	 *
	 * @returns An object containing the presence and results of the 'studiocms:astro:config' and 'studiocms:config:setup' hooks.
	 * - `astroConfig`: Indicates if the 'studiocms:astro:config' hook exists and provides its execution results.
	 * - `studiocmsConfig`: Indicates if the 'studiocms:config:setup' hook exists and provides its execution results.
	 */
	public async getHookResults(): Promise<PluginHookResults> {
		return {
			astroConfig: {
				hasHook: !!this.plugin.hooks['studiocms:astro:config'],
				hookResults: await this.runAstroConfigHook(),
			},
			studiocmsConfig: {
				hasHook: !!this.plugin.hooks['studiocms:config:setup'],
				hookResults: await this.runStudioCMSConfigHook(),
			},
		};
	}
}
