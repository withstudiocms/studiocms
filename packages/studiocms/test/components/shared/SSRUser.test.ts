/// <reference types="astro/client" />
import { experimental_AstroContainer as AstroContainer } from 'astro/container';
import { describe, expect, test } from 'vitest';
import SSRUser from '../../../src/components/shared/SSRUser.astro';
import { cleanAstroAttributes, MockAstroLocals } from '../../test-utils';

describe('SSRUser component', () => {
	test('render component', async () => {
		const container = await AstroContainer.create();
		const result = await container.renderToString(SSRUser, {
			locals: MockAstroLocals(),
			props: { name: 'mock', description: 'mock-admin' },
		});
		const cleanResult = cleanAstroAttributes(result, '/mock/path/SSRUser.astro');
		expect(cleanResult).toMatchSnapshot();
	});
});
