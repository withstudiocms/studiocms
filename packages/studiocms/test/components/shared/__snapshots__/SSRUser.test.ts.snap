// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SSRUser component > render component 1`] = `"<div class="sui-user-container"> <div class="sui-avatar-container"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" role="img" aria-label="Placeholder avatar for mock" height="24" width="24" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M15.75 6a3.75 3.75 0 1 1-7.5 0a3.75 3.75 0 0 1 7.5 0M4.501 20.118a7.5 7.5 0 0 1 14.998 0A17.9 17.9 0 0 1 12 21.75c-2.676 0-5.216-.584-7.499-1.632"/></svg> </div> <div class="sui-text-content"> <span class="sui-name">mock</span> <span class="sui-description">mock-admin</span> </div> </div> "`;
